hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -1794,6 +1791,65 @@ case class TableKeyBy(child: TableIR, keys: IndexedSeq[String], isSorted: Boolea
   }
 }
 
+/**
+ * Generate a table from the elementwise application of a body IR to a stream of `contexts`.
+ *
+ * @param contexts IR of type TStream[Any] whose elements are downwardly exposed to `body` as `cname`.
+ * @param globals  IR of type TStruct, downwardly exposed to `body` as `gname`.
+ * @param cname    Name of free variable in `body` referencing elements of `contexts`.
+ * @param gname    Name of free variable in `body` referencing `globals`.
+ * @param body     IR of type TStream[TStruct] that generates the rows of the table for each
+ *                 element in `contexts`, optionally referencing free variables Ref(cname) and
+ *                 Ref(gname).
+ * @param partitioner
+ * @param errorId  Identifier tracing location in Python source that created this node
+ */
+case class TableGen(contexts: IR,
+                    globals: IR,
+                    cname: String,
+                    gname: String,
+                    body: IR,
+                    partitioner: RVDPartitioner,
+                    errorId: Int = ErrorIDs.NO_ERROR
+                   ) extends TableIR {
+  private def typeCheck[A <: Type](varname: String, typ: Type)(implicit tag: ClassTag[A]): Option[A] =
```

Comment:
I disagree. I hate the errors from class cast exceptions (along with pattern match failures), especially when these happen in a delayed execution context (like `StackFrame`). More information is always better.

Reply:
The options are also allocated from the cast. I'm not sure how `asInstanceOf` is implemented, but a pattern match generates an `Option` (via `ClassTag.unapply`) anyway

Reply:
Rewritten without allocs.

Reply:
What's the additional information here? I'm contending that this is a much longer way of writing an exception that doesn't offer any more context than the stack trace and mismatched classes in the default classcastexception will give.

Reply:
The assertion contains the name of the offending argument. This allows me to test that the right assertion has been encountered.
Following our discussion, I tried moving this into `TypeCheck` but the problem is the `typ` is requested when computing the bindings used by the type checker.
I want type errors to arise in a predicable place in a manner that's easily testable so I'm going to keep a slimmed down version of this so that `TableGen` will be type correct by construction.